xml/registry.rnc

```diff
@@ -419,6 +419,9 @@ Extensions = element extensions {
 #       extension
 #   supported - profile name(s) supporting this extension, e.g. 'vulkan'
 #       or 'disabled' to never generate output.
+#   precedes - Vulkan version or name of extension that succeeded this
```

Comment:
I think it'd make more sense to tag the "promotion level" with the actual tag name, rather than relying on string manipulation outside of the xml structure. So something like "promotedby=" for things that are fully promoted, and then other terms can be introduceds, something like "succeededby" (or some less generic word) for things that are actively not working any more and have an alternative (e.g. AMD_negative_viewport_height) .

Reply:
Also I'm not sure if "full promotion" is meant to be "this completely overtakes all of the functionality of this extension" or just "it's in a major version now"?

I thought the former, but now re-reading it might be the latter... otherwise it seems 'promoted' should be on a lot more things, and I'm not sure why you even have it... is it just semantic tagging to say "this is a version"?

Reply:
Right, I didn't want to explode the attribute count. Basically for no text manipulation I would need four, something: `promotedtoext`, `promotedtovk`, `obsoletedbyext`, and `obsoletedbyvk`. Which also complicates the rules (you can use only one of those attributes on a tag).

BTW, how should you prefer I write multi-word attribute names? Just all lowercase as above?

<hr />

So "promotion" I define:

If extension **A** is promoted to extension or Vulkan version **B**, that means **B** has all the interfaces (commands, structures, enums, etc) **A** has with only name change (e.g. removing KHR, or EXT->KHR suffix change). Changes to behavior of **B**  must be backwards-compatible with **A** (i.e. App that switches from **A to **B**, and only S&R the names must work). Changes to enabling mechanism is tolerated (e.g. from `ppEnabledExtensions` to `pEnabledFeatures`).

So basically what happened with some extensions and Vulkan 1.1 +- the last part, and that it can happen between extensions (e.g. a promotion from EXT to KHR extension).

The generic "succession" would be:

If extension **A** is suceeded by extension or Vulkan version **B**, that means an App can achieve the same results or use case enabled by **A** by using **B** instead.
E.g. `debug_report` and `debug_utils` serve the same purpose, but do not share interfaces. `debug_utils` is just supposed to be newer (and better) than `debug_report`, so apps should prefer to use`debug_utils` (if they have a choice).

Reply:
Hopefully correctly rebased.

Ok, I see what the confusion is. I marked all the extensions as "succession" only. I was probably lazy to check which are full-fledged promotion, and never got back to it. The script should be ready for that case though...

Reply:
> Right, I didn't want to explode the attribute count. Basically for no text manipulation I would need four, something like: promotedtoext, promotedtovk, obsoletedbyext, and obsoletedbyvk. Which also complicates the rules (you can use only one of those attributes on a tag).

I think just "promotedby" and "obsoletedby" would be sufficient - then just detect whether the value is a version or an extension name. That seems like it'd be more sane IMO?

> BTW, how should you prefer I write multi-word attribute names? Just all lowercase as above?

Yea I think that's consistent.

> So, "promotion" I define:

> If extension A is promoted to extension or Vulkan version B, that means B has all the interfaces (commands, structures, enums, etc) A has with only name change (e.g. removing KHR, or EXT->KHR suffix change). Changes to behavior of B must be backwards-compatible with A (i.e. App that switches from A to B, and only S&R the names must work). Changes to enabling mechanism is tolerated (e.g. from ppEnabledExtensions to pEnabledFeatures).

That's mostly consistent with what we've defined internally, though we're also including renames of APIs and other compatible tweaks (e.g. adding additional parameters).

> The more generic "succession" would be:

> If extension A is suceeded by extension or Vulkan version B, that means an App can achieve the same results or use case enabled by A by using B instead.

I had to double check, but we've used the term _deprecated_ for this in the spec (see the AMD_negative_viewport_height appendix). So "deprecatedby" would be appropriate.

> E.g. debug_report and debug_utils serve the same purpose, but do not share interfaces. debug_utils is just supposed to be newer (and better) than debug_report, so apps should prefer to usedebug_utils (if they have a choice).

As per my comment above, I definitely think we'd consider this promotion. We should definitely clarify this in the spec; post 1.1 we have an aliasing mechanism for "identical" APIs with different names/suffixes, but haven't used this for debug utils since that came before the mechanism - we should probably retroactively do that actually... I'll look at raising a separate issue for that though.

> Grey area things like AMD_negative_viewport_height could also belong here. The spec claims "Due to some slight behavioral differences", so it cannot be a "promotion".

It's deprecated, as per the spec's current language.

Reply:
OK, I shall update the language.

ad `debug_report` and `debug_utils`:
AIS above, I did not check the extensions thoroughly, if they fit the "promotion" category, but I think it is outside it because of `VkDebugReportObjectTypeEXT` and `PFN_vkDebugUtilsMessengerCallbackEXT` being different and such.

As long as I am adding things to the XML, I would like at least one of the new attributes to not be too fuzzy, and so potentially usable by various scripts. Having two fuzzy attributes with several undocumented exceptions would defeat the purpose, and we could just have one. The "promoted" should be reserved for cases of almost verbatim duplicates (as was done with extensions assimilated by 1.1).